src/compute/lower.rs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines kernel to extract a lower case of a \[Large\]StringArray
+
+use crate::array::*;
+use crate::{
+    datatypes::DataType,
+    error::{ArrowError, Result},
+};
+
+fn utf8_lower<O: Offset>(array: &Utf8Array<O>) -> Utf8Array<O> {
+    let iter = array.values_iter().map(str::to_lowercase);
+
+    let new = Utf8Array::<O>::from_trusted_len_values_iter(iter);
+    new.with_validity(array.validity().cloned())
+}
+
+/// Returns an ArrayRef with lowercase of each of the elements in `array`.
+/// this function errors when the passed array is not a \[Large\]String array.
+pub fn lower(array: &dyn Array) -> Result<Box<dyn Array>> {
+    match array.data_type() {
+        // For binary and large binary, lower is no-op.
+        DataType::Binary | DataType::LargeBinary => unsafe {
```

Comment:
Could you clarify why we should support "lower" for Binary? Is there a well-defined concept of "lowercase" for blob/binary?

Reply:
Well, it's a decision that should be made by the maintainer.

There are three statuses:

- utf8 / large utf8: well as expected.
- binary / large binary: no-op, but no error.
- other data types: incorrect, and we should return an error.

Most implementations will allow users to lower a binary (even it is meaningless). If arrow2 doesn't support binary, we will need to check the datatype before calling `lower`.

Reply:
IMO, an implementation of Arrow should decide if a lower in binary makes sense. I don't think it does and I feel that arrow2 should error/panic.

Reply:
Addressed.